osu.Game/Online/Rooms/BeatmapAvailability.cs

```diff
@@ -0,0 +1,56 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System;
+
+namespace osu.Game.Online.Rooms
+{
+    /// <summary>
+    /// The local availability information about a certain beatmap for the client.
+    /// </summary>
+    public readonly struct BeatmapAvailability : IEquatable<BeatmapAvailability>
+    {
+        /// <summary>
+        /// The beatmap's availability state.
+        /// </summary>
+        public readonly DownloadState State;
+
+        /// <summary>
+        /// The beatmap's downloading progress, null when not in <see cref="DownloadState.Downloading"/> state.
+        /// </summary>
+        public readonly double? DownloadProgress;
+
+        /// <summary>
+        /// Constructs a new non-<see cref="DownloadState.Downloading"/> beatmap availability state.
+        /// </summary>
+        /// <param name="state">The beatmap availability state.</param>
+        /// <exception cref="ArgumentException">Throws if <see cref="DownloadState.Downloading"/> was specified in this constructor, as it has its own constructor (see <see cref="BeatmapAvailability(DownloadState, double)"/>.</exception>
+        public BeatmapAvailability(DownloadState state)
+        {
+            if (state == DownloadState.Downloading)
+                throw new ArgumentException($"{nameof(DownloadState.Downloading)} state has its own constructor, use it instead.");
+
+            State = state;
+            DownloadProgress = null;
+        }
```

Comment:
This is all sorts of wrong from an API design standpoint. If you want to do this, do it like this or similarly:

```csharp
private BeatmapAvailability()
{
}

public static BeatmapAvailability NotDownloaded() => new BeatmapAvailability
{
	State = NotDownloaded
};

public static BeatmapAvailability DownloadProgress(double progress) => new BeatmapAvailability
{
	State = Downloading,
	DownloadProgress = progress
};
```

Notice how you *can't* misuse the above without messing with this class.

Reply:
Just wasn't more creative you could say /shrug